### PR TITLE
JENKINS-53284 Fix warning message for changing workspace directories

### PR DESCRIPTION
**Pull request by charanbir**
See [JENKINS-53284](https://issues.jenkins-ci.org/browse/JENKINS-53284).

### Proposed changelog entries

* [JENKINS-53284](https://issues.jenkins-ci.org/browse/JENKINS-53284): only warn about changing workspace or builds directories on non fresh startups

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### CODE REVIEW

**Comment by charanbir** [comment]
This is my first open source contribution commit ever. Please let me know if I missed anything. Thank you.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: NB: can be written more concisely as
```
Level level = workspaceDir.equals(newWorkspacesDir) ? Level.WARNING : Level.INFO;
```

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: there is something wrong with the indentation here.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: won't this cause the WARNING message to always be logged on all startups except the first one?

**Comment by jtnord** [comment]
Hi @charanbir, 
The associated unit test (jenkins.model.JenkinsBuildsAndWorkspacesDirectoriesTest) is failing, possibly due to https://github.com/jenkinsci/jenkins/pull/3728#discussion_r232275121?

**Comment by charanbir** [comment]
Thank you for your input and guidance. I will make the necessary changes.

**Comment by charanbir** [comment]
@jtnord Changes made

**Comment by jtnord** [comment]
@charanbir Can I suggest you write a unit test for the expected behaviour, there is something not quite right with your code which will become obvious if you have a unit test.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (batmat) [inline core/src/main/java/jenkins/model/Jenkins.java]: I think this will not work. If the user changes it once, then again, the warning will not be shown when it should be.

I agree with James that writing a test for this is important here. This will make it obvious that it works or not, and will make much easier for you @charanbir to test and iterate on making sure your change does work.

**Review by batmat — CHANGES_REQUESTED** [review]
Awesome work! Thanks a lot for chiming in @charanbir !

We're almost there :smile_cat:. Please see to add an automated test, likely a new case in the existing `JenkinsBuildsAndWorkspacesDirectoriesTest`.
You will see a `LoggerRule` is already in use there with example on how to check if a log was issued or not.

Thanks again!

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: when will this be set to `INFO` and when will it be set to `WARNING`, what is the test condition in the surrounding block?

File: test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java
Comment (batmat) [inline test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java]: `logWasFoundAtLevel` would probably better reflect what it does.
Also, I think `logWasFound` should delegate to it.

File: test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java
Comment (batmat) [inline test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java]: For lambda clarity/declarativeness, I'd recommend putting two filters instead of using `&&`:

```java
.filter(record -> record.getMessage().contains(searched))
.filter(record -> record.getLevel().equals(level))
```

**Review by batmat — COMMENTED** [review]
I built that PR locally and tested it. For me it does not work as expected or I'm missing something.
The following use case issues the log as `INFO` when it should be a WARNING:

```
# move under a fresh empty directory
 # no sysprop specified
JENKINS_HOME=$PWD java -jar /path/to/jenkins.war --httpPort=8080
...
 # Following startup logs as `INFO` when it should be a WARNING
JENKINS_HOME=$PWD java -Djenkins.model.Jenkins.workspacesDir=bluh -jar /path/to/jenkins.war --httpPort=8080
...
 # Following startup logs as `WARNING` correctly, since bluh was changed to blah \o/
JENKINS_HOME=$PWD java -Djenkins.model.Jenkins.workspacesDir=blah -jar /path/to/jenkins.war --httpPort=8080
```

Also, seems like the `buildsDir` sysprop is not handled in this PR, but definitely fine by me to do it in a dedicated PR after this one gets merged.

Thanks a lot for your efforts @charanbir !

**Comment by charanbir** [comment]
Alright, I'll try to fix it. Please bear with me.

**Comment by charanbir** [comment]
@batmat 

- Thank you for reviewing. I made the changes you suggested, including the verbiage. 
- In all your commands you have initialized JENKINS_HOME yourself, so that value will never be null.
- I couldn't use the 2 variables `newWorkspacesDir` and `workspaceDir` to determine if it was a fresh startup so I created a boolean variable. This works and I tested all cases mentioned in the above comments. I also used the same logic for `buildsDir`.
- Please let me know if there are any concerns or changes.

**Comment by batmat** [comment]
@charanbir sorry for the lag, I've been busy on a few other things. I didn't forget this and hope to get to it again very soon. Thanks a lot for the work!

File: test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java
Comment (oleg-nenashev) [inline test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java]: tabs vs. spaces

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
`freshStartup` will not work correctly if you keep it in `config.xml` like it is done in this pull request:

* It is NOT guaranteed that fresh-start instances have no job data added in JENKINS_HOME
* It will suppress the message on the first upgrade to the version

Since the warning is still present as INFO, I request changes only for the data storage. It should be rather a separate marker file IMO

File: test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java
Comment (batmat) [inline test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java]: NIT: `logWasFound` body above should probably be rewritten to `logWasFoundAtLevel(searched, Level.ALL)`

File: test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java
Comment (batmat) [inline test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java]: NIT: see `anyMatch()` above. It's somehow unimportant here in test context, but it **is** important in general. Because what the code here is doing necessitates to go computing the whole list size. 

I.e. if `Collectors.toList()` is actually going to compute a list of billions of elements, it could take a lot of time, when anyMatch will just stop on the first match and yield the return result.

File: test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java
Comment (batmat) [inline test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java]: Your code formatter seems to have an issue :-)
image

**Review by batmat — APPROVED** [review]
LGTM from local testing.

**Comment by batmat** [comment]
@charanbir I've added the commits requested by Oleg myself, because I feel like I owe you some time given how long you've waited for this to be looked at again :-|. I hope you do not mind.

@oleg-nenashev I believe I addressed your concern, and used your recommendation to use a marker file instead. Care to have another look?

I've also retested locally, and this works as expected (first startup, `INFO` only, `WARN` next times if one changes the buildsDir or workspacesDir value -- note: this is/was anyway also JUnit tested).

File: test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java
Comment (batmat) [reply test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java]: my bad, ^ is wrong, because it's an exact level match. So it's not fully as straightforward as I was thinking.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (MRamonLeon) [inline core/src/main/java/jenkins/model/Jenkins.java]: Is this file ever removed? In a classic Jenkins instance it will always stay there after the first start. Doesn't it involves too much (adding an extra file for ever) for just a level in the warning? Isn't there any other way to catch we are in a fresh startup? I.e: the former build or workspace directories already exists or something like that.

File: test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java
Comment (MRamonLeon) [inline test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java]: Shouldn't you test the log is found at level INFO?

**Review by MRamonLeon — COMMENTED** [review]
Some doubts

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (batmat) [reply core/src/main/java/jenkins/model/Jenkins.java]: I believe 1) adding marker file is a simple enough solution and commonly used in Jenkins codebase (see the `secret.key.not-so-secret` file for instance and 2) this _could_ be useful for other code. Low-tech is sometimes also good too IMO: if something goes wrong, we can ask user to delete (or create) this file to trigger/revert to some other behavior.

File: test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java
Comment (batmat) [reply test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java]: At that stage, we have already started (i.e. **without** any flag), and about to restart), so in that first case: **no** log at all is expected, so this way works fine.

As you can see below, on the second restart after having set the property: then yes we must check the level more specifically.

Does that clarify @MRamonLeon?

**Comment by batmat** [comment]
@jtnord care to have a second look? Thanks!

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (MRamonLeon) [reply core/src/main/java/jenkins/model/Jenkins.java]: I agree low-tech is good, almost always. I was just asking if there were any other approach, with the existing infra (files, directories, status files)

File: test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java
Comment (MRamonLeon) [reply test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java]: Yes, I was asking to check when the message is raised in INFO level.

**Review by MRamonLeon — APPROVED** [review]
Anyway, LGTM

**Comment by batmat** [comment]
Retriggered a build, looked like an infra issue
```
Creating placeholder flownodes because failed loading originals.

GitHub has been notified of this commit’s build result

java.io.IOException: Tried to load head FlowNodes for execution Owner[Core/jenkins/PR-3728/36:Core/jenkins/PR-3728 #36] but FlowNode was not found in storage for head id:FlowNodeId 1:196
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.initializeStorage(CpsFlowExecution.java:678)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.onLoad(CpsFlowExecution.java:715)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.getExecution(WorkflowRun.java:664)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.onLoad(WorkflowRun.java:526)
	at hudson.model.RunMap.retrieve(RunMap.java:225)
	at hudson.model.RunMap.retrieve(RunMap.java:57)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.load(AbstractLazyLoadRunMap.java:501)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.load(AbstractLazyLoadRunMap.java:483)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.getByNumber(AbstractLazyLoadRunMap.java:381)
	at hudson.model.RunMap.getById(RunMap.java:205)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun$Owner.run(WorkflowRun.java:901)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun$Owner.get(WorkflowRun.java:912)
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext.getFlowExecution(CpsStepContext.java:474)
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext.getThreadGroupSynchronously(CpsStepContext.java:243)
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext.access$000(CpsStepContext.java:97)
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext$1.call(CpsStepContext.java:264)
	at org.jenkinsci.plugins.workflow.cps.CpsStepContext$1.call(CpsStepContext.java:262)
	at jenkins.util.ContextResettingExecutorService$2.call(ContextResettingExecutorService.java:46)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Finished: FAILURE
```

**Review by oleg-nenashev — APPROVED** [review]
Thanks to @batmat for addressing the change request

**Comment by batmat** [comment]
I consider this ready for merge. 
I will likely merge it later today or tomorrow (if the PR build goes green, obviously), to possibly wait for more feedback (even if quite unlikely given this has been here for quite long already :-)).

Thanks!

**Comment by jglick** [comment]
Rather than suppressing warnings, I would suggest just removing support for the old locations, which is long obsolete. As I noted in [JENKINS-21942](https://issues.jenkins-ci.org/browse/JENKINS-21942), contrary to expectations the old locations are used in _new_ installations under some circumstances.

**Comment by daniel-beck** [comment]
@charanbir Please do not delete the pull request template.
@batmat Please restore the PR template in such cases, or don't merge nontrivial PRs without is.

**Comment by daniel-beck** [comment]
Does this actually work as intended? The idea is to print a warning when the directory changes, but all this does is to make the first startup with any directory defined print a warning, and none of the subsequent startups, i.e. if the specified directory is changed between subsequent startups, no warning will be written.

**Comment by batmat** [comment]
> Please restore the PR template in such cases, or don't merge nontrivial PRs without is.

@daniel-beck my bad. I generally do it, but in this case, seems like I missed, maybe because the PR had been running for quite long.

**Comment by batmat** [comment]
> Does this actually work as intended?

AFAIR from my testing last week, yes. And there are automated tests also, so yes I think this works. 
I will anyway be able to confirm it soon: once the weekly is out, I plan to update Evergreen with this. I'll get back here if this actually does not work. 

I was even planning to use the Incremental release right away, but this is actually currently broken, so I had to wait for the weekly to deploy it :disappointed:. ([INFRA-1964](https://issues.jenkins-ci.org/browse/INFRA-1964))

**Pull request merged**
